Sources/Datadog/DatadogCore/Context/ContextValueReader.swift

```diff
@@ -0,0 +1,129 @@
+ */
+
+import Foundation
+
+/// Declares that a type can read values on demand.
+///
+/// A reader delivers elements to the receiver callback synchronously.
+/// The receiver's ``ContextValueReceiver/Value`` generic type must match the
+/// ``ContextValueReader/Value`` types declared by the reader.
+///
+/// The reader implements the ``ContextValuePublisher/read(_:)`` method
+/// to read the value and call the receiver.
+internal protocol ContextValueReader {
+    /// The kind of values this reader reads.
+    associatedtype Value
+
+    /// The initial value of the reader.
+    var initialValue: Value { get }
+
+    /// Reads the value synchronously and calls the **nonescaping** receiver.
+    ///
+    /// - Parameter receiver: The receiver to mutate.
+    func read(to receiver: inout Value)
+}
+
+// MARK: - Type-Erasure
+
+/// A reader that performs type erasure by wrapping another reader.
+///
+/// ``AnyContextValueReader`` is a concrete implementation of ``ContextValueReader``
+/// that has no significant properties of its own, and passes through values from its upstream
+/// reader.
+///
+/// Use ``AnyContextValueReader`` to wrap a reader whose type has details
+/// you don’t want to expose across API boundaries, such as different modules
+///
+/// You can use extension method ``ContextValueReader/eraseToAnyReader()``
+/// operator to wrap a publisher with ``ContextValueReader``.
+internal struct AnyContextValueReader<Value>: ContextValueReader {
+    /// The initial value of the reader.
+    let initialValue: Value
+
+    private var readBlock: (inout Value) -> Void
+
+    /// Creates a type-erasing reader to wrap the provided reader.
+    ///
+    /// - Parameter reader: A reader to wrap with a type-eraser.
+    init<Reader>(_ reader: Reader) where Reader: ContextValueReader, Reader.Value == Value {
+        self.initialValue = reader.initialValue
+        self.readBlock = reader.read
+    }
+
+    /// Creates a type-erasing reader to read value from a closure.
+    /// - Parameters:
+    ///   - initialValue: The initial value of the reader.
+    ///   - block: A block for reading the value.
+    init(initialValue: Value, read block: @escaping (inout Value) -> Void = { _ in }) {
+        self.initialValue = initialValue
+        self.readBlock = block
+    }
+
+    /// Appends a read operation and set the value's property at a given `keyPath`.
+    ///
+    /// - Parameters:
+    ///   - reader: The reader to append.
+    ///   - keyPath: The value's writable `keyPath`.
+    mutating func append<Reader>(reader: Reader, receiver keyPath: WritableKeyPath<Value, Reader.Value>) where Reader: ContextValueReader {
+        let previousRead = readBlock
+        // perform recursive read
+        readBlock = { value in
+            // first perform the previous read operation
+            previousRead(&value)
+            // then read the property
+            reader.read(to: &value[keyPath: keyPath])
+        }
```

Comment:
Here, I'm using recursion to chain reads for value properties. Another option would be to create another `ContextValueReader` concrete type that keeps an array of read operations, such as:
```swift
internal struct KeyPathContextValueReader<Value>: ContextValueReader {
    let initialValue: Value

    private var reads: [(inout Value) -> Void] = []

    init(initialValue: Value) {
        self.initialValue = initialValue
    }

    mutating func append<Reader>(reader: Reader, receiver keyPath: WritableKeyPath<Value, Reader.Value>) where Reader: ContextValueReader {
        reads.append({ value in
            reader.read(to: &value[keyPath: keyPath])
        })
    }

    func read(to receiver: inout Value) {
        reads.forEach { read in
            read(&receiver)
        }
    }
}
```

Reply:
I went for the `KeyPathContextValueReader`.